src/eval_latency.py

```diff
@@ -0,0 +1,63 @@
+import json
+import sys
+import yaml
+import torch
+import time
+import random
+import subprocess
+
+from argparse import ArgumentParser
+from models import RetNetModelHF, TransformerModelHF
+from pathlib import Path
+from torchscale.architecture.config import RetNetConfig, DecoderConfig
+from transformers import AutoModel, AutoConfig, AutoModelForCausalLM, PreTrainedTokenizerFast
+from typing import Optional, List
+from utils import Struct
+from generate import generate_text_length_n
+from models import RetNetModel, TransformerModel
+
+# measure the time to generate a single token
+def eval_single_token_generation_time(config: Struct, input_tokens: torch.Tensor):
+    generation_time = generate_text_length_n(config=config, n=1, input_tokens=input_tokens)
+    return generation_time
+
+# measure the time to generate n tokens
+def eval_n_tokens_generation_time(config: Struct, n_tokens_to_generate: int, input_tokens: torch.Tensor):
+    generation_time = generate_text_length_n(config=config, n=n_tokens_to_generate, input_tokens=input_tokens)
+    return generation_time
+
+def eval_latency(config: Struct):
+    test_sequence_lengths = [i * (config.seq_len // config.seq_len_frac) for i in range(1, config.seq_len_frac)]
+    test_sequence_lengths.append(config.seq_len - 1)
+    results = {}
+    random_tokens = [random.randint(0, config.vocab_size - 1) for _ in range(n_tokens)]
+    # the first couple token generation take longer than all the following generations
+    generate_text_length_n(config=config, n=1, input_tokens=random_tokens[0:1])
+    generate_text_length_n(config=config, n=1, input_tokens=random_tokens[0:1])
+
+    max_len_generated = False
+
+    for test_seq_len in test_sequence_lengths:
+        print('test_seq_len: ', test_seq_len)
+        iteration_results = {'single_token_times': [], 'n_token_times': []}
+        input_tokens = random_tokens[:test_seq_len]
+        print('input_tokens length: ', len(input_tokens))
+
+        print("Measuring single token generation time...")
+        for _ in range(config.num_latency_trials):
+            iteration_results["single_token_times"].append(eval_single_token_generation_time(config, input_tokens))
+
+        print(f"Measuring {test_seq_len} token generation time...")
+        if test_seq_len >= config.seq_len - config.n_tokens_to_generate and not max_len_generated:
+            for _ in range(config.num_latency_trials):
+                iteration_results["n_token_times"].append(eval_n_tokens_generation_time(config, config.n_tokens_to_generate, input_tokens[:config.seq_len - config.n_tokens_to_generate]))
+                max_len_generated = True
+        else:
```

Comment:
A comment explaining what cases this if-else aims to handle would be nice. More comments throughout this whole python file would be nice, actually. The print statements, type hinting, and variable names help, but comments are extra nice.